q2_ms/xcms/read_ms_experiment.py

```diff
@@ -0,0 +1,97 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2025, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+import copy
+import os
+import tempfile
+
+from qiime2 import Metadata
+
+from q2_ms.types import XCMSExperimentDirFmt, mzMLDirFmt
+from q2_ms.utils import run_r_script
+
+
+def read_ms_experiment(
+    spectra: mzMLDirFmt,
+    sample_metadata: Metadata = None,
+) -> XCMSExperimentDirFmt:
+    # Create parameters dict
+    params = copy.copy(locals())
+
+    # Init XCMSExperimentDirFmt
+    xcms_experiment = XCMSExperimentDirFmt()
+
+    # Add output path to params
+    params["output_path"] = str(xcms_experiment)
+
+    if sample_metadata is not None:
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            # Validate sample metadata IDs
+            sample_metadata_table = sample_metadata.to_dataframe()
+            _validate_metadata(sample_metadata_table, str(spectra))
+
+            # Save sample metadata to tsv and add to params
+            tsv_path = os.path.join(tmp_dir, "sample_metadata.tsv")
+            sample_metadata_table.to_csv(tsv_path, sep="\t")
+            params["sample_metadata"] = tsv_path
+
+            # Run R script
+            run_r_script(params, "read_ms_experiment", "XCMS")
+    else:
+        run_r_script(params, "read_ms_experiment", "XCMS")
```

Comment:
You can simplify this a bit:
```suggestion
    with tempfile.TemporaryDirectory() as tmp_dir:
        if sample_metadata is not None:
            # Validate sample metadata IDs
            sample_metadata_table = sample_metadata.to_dataframe()
            _validate_metadata(sample_metadata_table, str(spectra))

            # Save sample metadata to tsv and add to params
            tsv_path = os.path.join(tmp_dir, "sample_metadata.tsv")
            sample_metadata_table.to_csv(tsv_path, sep="\t")
            params["sample_metadata"] = tsv_path

        # Run R script
        run_r_script(params, "read_ms_experiment", "XCMS")
```